In [1]:
i <- 1
while (i < 6) {
  print(i)
  i <- i + 1
  if (i == 4) {
    break
  }
}

[1] 1
[1] 2
[1] 3


In [17]:
fruits <- list("apple", "banana", "cherry")

for (x in fruits) {
  print(x)
}

[1] "apple"
[1] "banana"
[1] "cherry"


In [18]:
dice <- c(1, 2, 3, 4, 5, 6)

for (x in dice) {
  print(x)
}

[1] 1
[1] 2
[1] 3
[1] 4
[1] 5
[1] 6


In [19]:
fruits <- list("apple", "banana", "cherry")

for (x in fruits) {
  if (x == "cherry") {
    break
  }
  print(x)
}

[1] "apple"
[1] "banana"


Next

With the next statement, we can skip an iteration without terminating the loop:

In [20]:
fruits <- list("apple", "banana", "cherry")

for (x in fruits) {
  if (x == "banana") {
    next
  }
  print(x)
}

[1] "apple"
[1] "cherry"


In [21]:
adj <- list("red", "big", "tasty")

fruits <- list("apple", "banana", "cherry")
  for (x in adj) {
    for (y in fruits) {
      print(paste(x, y))
  }
}

[1] "red apple"
[1] "red banana"
[1] "red cherry"
[1] "big apple"
[1] "big banana"
[1] "big cherry"
[1] "tasty apple"
[1] "tasty banana"
[1] "tasty cherry"


In [2]:
my_function <- function(fname) {
  paste(fname, "Griffin")
}

my_function("Peter")
my_function("Lois")
my_function("Stewie")

[1] "Peter Griffin"

[1] "Lois Griffin"

[1] "Stewie Griffin"

Number of argument

In [10]:
my_function <- function(fname, lname) {
  paste(fname, lname)
}

my_function("Peter", "Griffin")

[1] "Peter Griffin"

Default Parameter Value

***Nested function***

In [12]:
my_function <- function(country = "Norway") {
  paste("I am from", country)
}

my_function("Sweden")
my_function("India")
my_function() # will get the default value, which is Norway
my_function("USA")

[1] "I am from Sweden"

[1] "I am from India"

[1] "I am from Norway"

[1] "I am from USA"

return values

In [13]:
my_function <- function(x) {
  return (5 * x)
}

print(my_function(3))
print(my_function(5))
print(my_function(9))

[1] 15
[1] 25
[1] 45


Example Explained


The function tells x to add y.

The first input Nested_function(2,2) is "x" of the main function.

The second input Nested_function(3,3) is "y" of the main function.

The output is therefore (2+2) + (3+3) = 10.

In [3]:
Nested_function <- function(x, y) {
  a <- x + y
  return(a)
}

print (Nested_function(2,2))

print (Nested_function(3,3))

print (Nested_function(4,6))

Nested_function(Nested_function(2,2), Nested_function(3,3))

[1] 4
[1] 6
[1] 10


[1] 10

Global variables

In [7]:
txt <- "global variable"
my_function <- function() {
  txt = "fantastic"
  paste("R is", txt)
}

my_function()

txt # print txt

[1] "R is fantastic"

[1] "global variable"

The Global Assignment Operator

In [8]:
my_function <- function() {
txt <<- "fantastic"
  paste("R is", txt)
}

my_function()

print(txt)

[1] "R is fantastic"

[1] "fantastic"


To change the value of a global variable inside a function, refer to the variable by using the global assignment operator <<-:

In [9]:
txt <- "awesome"
my_function <- function() {
  txt <<- "fantastic"
  paste("R is", txt)
}

my_function()

paste("R is", txt)

[1] "R is fantastic"

[1] "R is fantastic"

***Recursion***


In this example, tri_recursion() is a function that we have defined to call itself ("recurse"). We use the k variable as the data, which decrements (-1) every time we recurse. The recursion ends when the condition is not greater than 0 (i.e. when it is 0).

In [16]:
tri_recursion <- function(k) {
  if (k > 0) {
    paste ("k is ", k)
    result <- k + tri_recursion(k - 1)
    print(result)
  } else {
    result = 0
    return(result)
  }
}
tri_recursion(6)

[1] 1
[1] 3
[1] 6
[1] 10
[1] 15
[1] 21


In [17]:
paste ("k is ", 3)

[1] "k is  3"

***Data Set***

There is a popular built-in data set in R called "mtcars" (Motor Trend Car Road Tests), which is retrieved from the 1974 Motor Trend US Magazine.

In the examples below (and for the next chapters), we will use the mtcars data set, for statistical purposes:

In [1]:
mtcars

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

In [2]:
Data_Cars <- mtcars

use the question mark (?) to get information about the mtcars data set:

In [6]:
?mtcars

In [3]:
dim(Data_Cars)

[1] 32 11

Use the dim() function to find the dimensions of the data set, and the names() function to view the names of the variables:

In [4]:
names(Data_Cars)

[1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
[11] "carb"

Use the rownames() function to get the name of each row in the first column, which is the name of each car:

In [5]:
rownames(Data_Cars)

[1] "Mazda RX4"           "Mazda RX4 Wag"       "Datsun 710"         
 [4] "Hornet 4 Drive"      "Hornet Sportabout"   "Valiant"            
 [7] "Duster 360"          "Merc 240D"           "Merc 230"           
[10] "Merc 280"            "Merc 280C"           "Merc 450SE"         
[13] "Merc 450SL"          "Merc 450SLC"         "Cadillac Fleetwood" 
[16] "Lincoln Continental" "Chrysler Imperial"   "Fiat 128"           
[19] "Honda Civic"         "Toyota Corolla"      "Toyota Corona"      
[22] "Dodge Challenger"    "AMC Javelin"         "Camaro Z28"         
[25] "Pontiac Firebird"    "Fiat X1-9"           "Porsche 914-2"      
[28] "Lotus Europa"        "Ford Pantera L"      "Ferrari Dino"       
[31] "Maserati Bora"       "Volvo 142E"

If you want to print all values that belong to a variable, access the data frame by using the $ sign, and the name of the variable (for example cyl (cylinders)):

In [7]:
Data_Cars$cyl

[1] 6 6 4 6 8 6 8 4 4 6 6 8 8 8 8 8 8 4 4 4 4 8 8 8 8 4 4 4 8 6 8 4

min() and max() functions can be used to find the lowest or highest value in a set:

In [9]:
Data_Cars <- mtcars

Data_Cars$hp

[1] 110 110  93 110 175 105 245  62  95 123 123 180 180 180 205 215 230  66  52
[20]  65  97 150 150 245 175  66  91 113 264 175 335 109

In [10]:
max(Data_Cars$hp)
min(Data_Cars$hp)

[1] 335

[1] 52

Now we know that the largest horsepower value in the set is 335, and the lowest 52.

We could take a look at the data set and try to find out which car these two values belongs to:

In [11]:
Data_Cars

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

By observing the table, it looks like the largest hp value belongs to a Maserati Bora, and the lowest belongs to a Honda Civic.

However, it is much easier (and safer) to let R find out this for us.

For example, we can use the which.max() and which.min() functions to find the index position of the max and min value in the table:

In [12]:
Data_Cars <- mtcars

which.max(Data_Cars$hp)
which.min(Data_Cars$hp)

[1] 31

[1] 19

Or even better, combine which.max() and which.min() with the rownames() function to get the name of the car with the largest and smallest horsepower:

In [13]:
Data_Cars <- mtcars

rownames(Data_Cars)[which.max(Data_Cars$hp)]
rownames(Data_Cars)[which.min(Data_Cars$hp)]

[1] "Maserati Bora"

[1] "Honda Civic"

Now we know for sure:
Maserati Bora is the car with the highest horsepower, and Honda Civic is the car with the lowest horsepower.

Mean, Median, and Mode

In statistics, there are often three values that interests us:

Mean - The average value
Median - The middle value
Mode - The most common value

In [14]:
Data_Cars <- mtcars

mean(Data_Cars$wt)

[1] 3.21725

In [15]:
Data_Cars <- mtcars

median(Data_Cars$wt)

[1] 3.325

In [16]:
summary(Data_Cars)

      mpg             cyl             disp             hp       
 Min.   :10.40   Min.   :4.000   Min.   : 71.1   Min.   : 52.0  
 1st Qu.:15.43   1st Qu.:4.000   1st Qu.:120.8   1st Qu.: 96.5  
 Median :19.20   Median :6.000   Median :196.3   Median :123.0  
 Mean   :20.09   Mean   :6.188   Mean   :230.7   Mean   :146.7  
 3rd Qu.:22.80   3rd Qu.:8.000   3rd Qu.:326.0   3rd Qu.:180.0  
 Max.   :33.90   Max.   :8.000   Max.   :472.0   Max.   :335.0  
      drat             wt             qsec             vs        
 Min.   :2.760   Min.   :1.513   Min.   :14.50   Min.   :0.0000  
 1st Qu.:3.080   1st Qu.:2.581   1st Qu.:16.89   1st Qu.:0.0000  
 Median :3.695   Median :3.325   Median :17.71   Median :0.0000  
 Mean   :3.597   Mean   :3.217   Mean   :17.85   Mean   :0.4375  
 3rd Qu.:3.920   3rd Qu.:3.610   3rd Qu.:18.90   3rd Qu.:1.0000  
 Max.   :4.930   Max.   :5.424   Max.   :22.90   Max.   :1.0000  
       am              gear            carb      
 Min.   :0.0000   Min.   :3.000  